### PR TITLE
VDS-2593 [Export] Export descriptions with selected filter

### PR DESCRIPTION
**Pull request by trueboroda**
Description: 
image

QA-test: VDS-2596

Demo-test: VDS-2597

Download artifact URL: https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.DescriptionExportImport_1.0.0-pr-12-936500e9.zip

### CODE REVIEW

**Review by Pav-el — APPROVED** [review]

File: src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs [hunk 1]
```diff
@@ -16,11 +17,25 @@ public class ExportController : ControllerBase
...
+        [HttpPost]
+        [Route("count")]
+        public async Task<ActionResult<ExportTotalCountResponse>> GetTotalCount([FromBody] ExportDataRequest request)
+        {
+            var criteria = request.ToSearchCriteria();
+            criteria.Take = 0;
+
+            var searchResult = await _productEditorialReviewSearchService.SearchEditorialReviewsAsync(criteria, deepSearch: true);
+
+            return Ok(new ExportTotalCountResponse { TotalCount = searchResult.TotalCount });
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs:38]: Maybe use anonymous object? Like 
```suggestion
            return Ok(new { TotalCount = searchResult.TotalCount });
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 2]
```diff
@@ -16,15 +19,23 @@ public class ProductEditorialReviewSearchService : IProductEditorialReviewSearch
...
+        public ProductEditorialReviewSearchService(Func<ICatalogRepository> catalogRepositoryFactory, IProductEditorialReviewService productEditorialReviewService,
+            IListEntrySearchService listEntrySearchService)
         {
             _catalogRepositoryFactory = catalogRepositoryFactory;
             _productEditorialReviewService = productEditorialReviewService;
+            _listEntrySearchService = listEntrySearchService;
         }
 
-        public async Task<ProductEditorialReviewSearchResult> SearchEditorialReviewsAsync(ProductEditorialReviewSearchCriteria criteria)
+        public async Task<ProductEditorialReviewSearchResult> SearchEditorialReviewsAsync(ProductEditorialReviewSearchCriteria criteria, bool deepSearch = false)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]: Maybe add deep search property to the search criteria?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 3]
```diff
@@ -36,9 +47,19 @@ public async Task<ProductEditorialReviewSearchResult> SearchEditorialReviewsAsyn
 
             descriptionQuery = descriptionQuery.Where(x => x.CatalogItem.CatalogId == criteria.CatalogId);
 
-            descriptionQuery = !string.IsNullOrEmpty(criteria.CategoryId) ?
-                descriptionQuery.Where(x => x.CatalogItem.CategoryId == criteria.CategoryId) :
-                descriptionQuery.Where(x => x.CatalogItem.CategoryId == null);
+            if (criteria.CategoryIds.Length > 0 && criteria.ItemIds.Length > 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            if (!criteria.CategoryIds.Length.IsNullOrEmpty() && !criteria.ItemIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 4]
```diff
@@ -36,9 +47,19 @@ public async Task<ProductEditorialReviewSearchResult> SearchEditorialReviewsAsyn
...
+            {
+                descriptionQuery = descriptionQuery.Where(x => criteria.CategoryIds.Contains(x.CatalogItem.CategoryId)
+                || criteria.ItemIds.Contains(x.ItemId));
+            }
+            else if (criteria.CategoryIds.Length > 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            else if (!criteria.CategoryIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 5]
```diff
@@ -36,9 +47,19 @@ public async Task<ProductEditorialReviewSearchResult> SearchEditorialReviewsAsyn
...
+            {
+                descriptionQuery = descriptionQuery.Where(x => criteria.CategoryIds.Contains(x.CatalogItem.CategoryId));
+            }
+            else if (criteria.ItemIds.Length > 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            else if (!criteria.ItemIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs (hunk 3 above)
Comment (pushnitsa) [reply src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]: For fixing possible NullReferenceExceprion if ItemIds or categoryIds are null

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 6]
```diff
@@ -73,5 +94,85 @@ protected virtual IList<SortInfo> BuildSortExpression(ProductEditorialReviewSear
...
+
+        /// <summary>
+        /// Extend categories criteria with children categories. For the deep searching.
+        /// </summary>
+        /// <param name="searchCriteria"></param>
+        /// <returns></returns>
+        private async Task ExtendSearchCriteriaForDeepSearch(ProductEditorialReviewSearchCriteria searchCriteria)
+        {
+            // All from catalog
+            if (string.IsNullOrEmpty(searchCriteria.Keyword) && searchCriteria.CategoryIds.Length == 0 && searchCriteria.ItemIds.Length == 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            if (string.IsNullOrEmpty(searchCriteria.Keyword) && searchCriteria.CategoryIds.IsNullOrEmpty() && searchCriteria.ItemIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 7]
```diff
@@ -73,5 +94,85 @@ protected virtual IList<SortInfo> BuildSortExpression(ProductEditorialReviewSear
...
+                    var productsIds = listEntries.Where(x => x.Type.EqualsInvariant("product")).Select(x => x.Id).ToArray();
+                    searchCriteria.CategoryIds = categoriesIds;
+                    searchCriteria.ItemIds = productsIds;
+
+                    return;
+                }
+            }
+
+            // All with search by keyword 
+            if (!string.IsNullOrEmpty(searchCriteria.Keyword) && searchCriteria.CategoryIds.Length == 0 && searchCriteria.ItemIds.Length == 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            if (!string.IsNullOrEmpty(searchCriteria.Keyword) && searchCriteria.CategoryIds.IsNullOrEmpty() && searchCriteria.ItemIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 8]
```diff
@@ -73,5 +94,85 @@ protected virtual IList<SortInfo> BuildSortExpression(ProductEditorialReviewSear
...
+                    var listEntries = listEntrySearchResult.Results;
+                    var categoriesIds = listEntries.Where(x => x.Type.EqualsInvariant("category")).Select(x => x.Id).ToArray();
+                    var productsIds = listEntries.Where(x => x.Type.EqualsInvariant("product")).Select(x => x.Id).ToArray();
+                    searchCriteria.CategoryIds = categoriesIds;
+                    searchCriteria.ItemIds = productsIds;
+                }
+            }
+
+            // Extend CategoryIds with children. In case with searching by keyword this will work also. 
+            if (searchCriteria.CategoryIds.Length > 0)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
            if (!searchCriteria.CategoryIds.IsNullOrEmpty())
```

File: src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs [hunk 9]
```diff
@@ -16,11 +17,25 @@ public class ExportController : ControllerBase
...
+        [HttpPost]
+        [Route("count")]
+        public async Task<ActionResult<object>> GetTotalCount([FromBody] ExportDataRequest request)
+        {
+            var criteria = request.ToSearchCriteria();
+            criteria.Take = 0;
+
+            var searchResult = await _productEditorialReviewSearchService.SearchEditorialReviewsAsync(criteria, deepSearch: true);
+
+            return Ok(new { searchResult.TotalCount });
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs]: Could you please replace the anonymous object by an integer?

File: src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs [hunk 10]
```diff
@@ -73,5 +94,85 @@ protected virtual IList<SortInfo> BuildSortExpression(ProductEditorialReviewSear
...
             return sortInfos;
         }
+
+
+        /// <summary>
+        /// Extend categories criteria with children categories. For the deep searching.
+        /// </summary>
+        /// <param name="searchCriteria"></param>
+        /// <returns></returns>
+        private async Task ExtendSearchCriteriaForDeepSearch(ProductEditorialReviewSearchCriteria searchCriteria)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DescriptionExportImportModule.Data/Services/ProductEditorialReviewSearchService.cs]:
```suggestion
        private async Task ExtendSearchCriteriaForDeepSearchAsync(ProductEditorialReviewSearchCriteria searchCriteria)
```

**Review by pushnitsa — APPROVED** [review]

File: src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs (hunk 9 above)
Comment (trueboroda) [reply src/VirtoCommerce.DescriptionExportImportModule.Web/Controllers/Api/ExportController.cs]: Angularjs $resource does not support response with a scalar value

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**